File: Java/src/com/company/Lists/LC148_sortList.java
package com.company.Lists;

public class LC148_sortList {
    public ListNode sortList(ListNode head) {
        /*
            top down mergesort solution
            O(nlogn) time
            O(logn) space
        */

        if(head == null || head.next == null) return head;
        ListNode fast = head;
        ListNode slow = head;
        while(fast.next != null && fast.next.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }
        ListNode mid = slow.next;
        slow.next = null;

        return merge(sortList(head), sortList(mid));

    }

    private ListNode merge(ListNode l1, ListNode l2){
        ListNode guard = new ListNode(-1);
        ListNode tail = guard;
        while(l1 != null && l2 != null){
            if(l1.val < l2.val){
                tail.next = l1;
                tail = tail.next;
                l1 = l1.next;
            }else{
                tail.next = l2;
                tail = tail.next;
                l2 = l2.next;
            }
        }

        if(l1 == null){
            tail.next = l2;
        }
        if(l2 == null){
            tail.next = l1;
        }
        return guard.next;

    }
//    public ListNode sortList(ListNode head) {
//        /*
//            bottom up mergesort
//            O(nlogn)
//            O(1)space
//         */
//    }
}

File: python/List/LC148_sortList.py
class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        # Top-down O(nLogn) time, O(logn) space for recursion
        if head == None or head.next == None:
            return head
        slow = head
        fast = head
        while(fast.next and fast.next.next):
            slow = slow.next
            fast = fast.next.next

        second = slow.next
        slow.next = None
        first_half = self.sortList(head)
        second_half = self.sortList(second)
        return self.merge(first_half, second_half)

    def merge(self, first, second):
        dummy = ListNode(-1)
        cur = dummy
        while(first or second):
            if first == None:
                cur.next = second
                second = None
            elif second == None:
                cur.next = first
                first = None

            elif first.val <= second.val:
                cur.next = first
                cur = cur.next
                first = first.next
            else:
                cur.next = second
                cur = cur.next
                second = second.next
        return dummy.next
